### ISSUE TITLE
[@odata2ts/odata2model] Cannot Generate Types for Functions with Collection Parameters

### ISSUE THREAD

**Issue by merlosy** [odata2ts/odata2ts#72]
## Bug report

Custom functions associated types are missing from the generated "Qxxx" file and the "xxxModel" file, even though they are references by the generated service (class).
It is then used in the [resource path of URL components](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_URLComponents) for odata v4.

Here is an extract of the xml file:
```xml
<edmx:Edmx xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx" Version="4.0">
  <edmx:DataServices>
    <Schema xmlns="http://docs.oasis-open.org/odata/ns/edm" Namespace="XEntities">
      [...]
    </Schema>
    <Schema xmlns="http://docs.oasis-open.org/odata/ns/edm" Namespace="Default">
        <Function Name="FacetFilter" IsBound="true">
          <Parameter Name="bindingParameter" Type="Collection(XEntities.AEntity)" />
          <Parameter Name="themeIds" Type="Collection(Edm.Guid)" Nullable="false">
            <Annotation Term="Org.OData.Core.V1.OptionalParameter" />
          </Parameter>
          <Parameter Name="categoryIds" Type="Collection(Edm.Guid)" Nullable="false">
            <Annotation Term="Org.OData.Core.V1.OptionalParameter" />
          </Parameter>
          <ReturnType Type="Collection(XEntities.AEntity)" />
        </Function>
        <Function Name="YYYFilter" IsBound="true">
          <Parameter Name="bindingParameter" Type="Collection(XEntities.AEntity)" />
          <Parameter Name="themeIds" Type="Collection(Edm.Guid)" Nullable="false">
            <Annotation Term="Org.OData.Core.V1.OptionalParameter" />
          </Parameter>
          <Parameter Name="categoryIds" Type="Collection(Edm.Guid)" Nullable="false">
            <Annotation Term="Org.OData.Core.V1.OptionalParameter" />
          </Parameter>
          <ReturnType Type="Collection(XEntities.YYYFilterEntity)" />
        </Function>
    </Schema>
  </edmx:DataServices>
</edmx:Edmx>
```

In the called URI it should look like: `.../aentities/FacetFilter(themeIds=[],%20categoryIds=[])?$skip=0&$top=30`

**Comment by merlosy** [comment]
I think in my case `QParam<Type extends string | number | boolean>` should need to manage a `string[]`. The current design seems to limit this

**Comment by texttechne** [comment]
Hi @merlosy,

while function generation works in general, you're right, that collection parameters are currently not supported. I've never had any examples for this. So first of all, thanks for the example. I couldn't even find any documentation in the spec.

A couple of questions:
- Do you know if collection parameters in any way supported by V2 as well?
- interesting for me: With which software do you generate OData?

Best Regards

**Comment by merlosy** [comment]
Hey!
I'm not the backend expert in the team, but I'll go with what I found out asking around.

Our backend is in C# based on .net 6 and we use a nuget package to handle odata:
```xml
<PackageReference Include="Microsoft.AspNetCore.OData" Version="8.0.11" />
```
The $metadata.xml file is auto generated based on the code. The definition of the odata function is done using the ODataModelBuilder class.
I can't really answer if this is supported in v2 as we went directly to v4 since the project is quite recent.

This is the part of the docs that briefly mentions it: http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_InvokingaFunction :
> Parameter values passed to functions MUST be specified either as a URL literal (for primitive values) or as a JSON formatted OData object (for complex values, or collections of primitive or complex values). Entity typed values are passed as JSON formatted entities that MAY include a subset of the properties, or just the entity reference, as appropriate to the function.  

There is a course on Pluralsight that seems to go into details. I'm not trying to promote the website but that's what the developer used to ramp up: https://www.pluralsight.com/courses/building-consistent-restful-api-odata-v4-asp-dot-net-core. Here is the repo associated with the course: https://github.com/lkroliko/Building-a-Consistent-RESTful-API-with-OData-V4-in-.NET5, and especially this line mentions a collection for odata function [Startup.cs#L107](https://github.com/lkroliko/Building-a-Consistent-RESTful-API-with-OData-V4-in-.NET5/blob/master/AirVinyl.WebApi/Startup.cs#L107)

I hope this helps.
Cheers!

**Comment by texttechne** [comment]
Has been solved in master.
Here's [a test](https://github.com/odata2ts/odata2ts/blob/main/packages/odata-query-objects/test/operation/QFunction.test.ts#L101).

**Comment by texttechne** [comment]
@merlosy this feature has been released. Feedback would be very welcome here.

**Comment by texttechne** [comment]
QFunction & QAction are not generated with QComplexParam => `new undefined(....`

**Comment by texttechne** [comment]
Hi @merlosy,

sorry, I've realized that I closed the ticket instead of you. I'm afraid that I was too occupied with having a clean list of issues 😅 

Some proper tests were missing regarding QObject generation, so that failures regarding generated params were not detected. I've added more tests and also an example project with "edge cases", which serves as integration test for the generated stuff. This fix has been released with `@odata2ts/odata2ts` version `0.18.2`.

It would be really cool, if you could test this feature against a living server. 

Best Regards,
Hubert
